src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -813,9 +813,12 @@ private void SetFailedResult(Status status)
         // 3. There is an existing activity (to enable activity propagation)
         if (diagnosticSourceEnabled || Logger.IsEnabled(LogLevel.Critical) || Activity.Current != null)
```

Comment:
The story is unfortunately a bit messy, but I'd suggest using the same logic that HttpClient uses to create the Activity:
https://source.dot.net/#System.Net.Http/System/Net/Http/DiagnosticsHandler.cs,48
https://source.dot.net/#System.Net.Http/System/Net/Http/DiagnosticsHandler.cs,56
I think the important difference is the fallback to use Activity constructor. ActivitySource.CreateActivity() will return a null Activity if no ActivityListener requests it, which would break cases 2,3, and 4 if there is no fallback.

There are probably four different scenarios that are interesting:
1. Folks using the new ActivityListener APIs want to be able to turn on the Activity just by subscribing via ActivityListener. As-is this would be insufficient because ActivityListener doesn't necessarily cause any of the three conditions the code checks for to be true.
2. Folks using the DiagnosticListener APIs to get notified. For newer tools hopefully the only use for this is if they want to populate custom tags on the Activity by using the ActivityStartData and ActivityStopData that is passed in the WriteDiagnosticEvent calls. Eventually I hope we can deprecate the check for DiagnosticListener.IsEnabled(GrpcDiagnostics.ActivityName), but we'd want to give plenty of time for tools to adopt the ActivityListener mechanism first.
3. ILogger can log Activity trace information (the Logger.IsEnabled(LogLevel.Critical) check). I hope in the future we can stop the practice of implicitly driving distributed tracing configuration via logging configuration, but we probably need something like https://github.com/dotnet/runtime/pull/90201 for distributed tracing first and we aren't there yet.
4. Propagating trace information from inbound to outbound networking requests even when none of 1-3 are activated (The Activity.Current != null check). Similar to the logging case, I hope this is one we can avoid the need for this in the future when we have more explicit configuration APIs.

Reply:
> I think the important difference is the fallback to use Activity constructor.

Would this still allow us to remove the Reflection usage in OpenTelemetry? I think we should eliminate this Reflection usage in OpenTelemetry when running on the latest bits.

https://github.com/open-telemetry/opentelemetry-dotnet/pull/4695#discussion_r1276692304

My understanding is the reason they are using Reflection today is because libraries are constructing the Activity through a constructor, and the Activity.Source and Activity.Kind properties are not being set (and they can't be set publicly - thus Reflection).